### DISCUSSION TITLE
Need some help with how to load a python function once and use it over and over instead of loading it every time

### DISCUSSION THREAD

**Discussion by vamshiaruru-virgodesigns** [PyO3/pyo3#2755] [Questions] [answered]
Hi, total noob to rust and pyo3 here. I am following the call python from rust documentation, and I have a few questions. Currently I have code like:
```rust
pub fn func_py(number: usize) -> PyResult<()> {
    Python::with_gil(|py| {
        let start = Instant::now();
        let builtins = PyModule::import(py, "builtins")?;
        let sum_func = builtins.getattr("sum")?;
        println!("Time taken to get function: {}", start.elapsed().as_nanos());
        let start = Instant::now();
        let total: i32 = sum_func.call1((vec![1, 2, number],))?.extract()?;
        println!("{total}, time taken {}", start.elapsed().as_nanos());
        Ok(())
    })
}
```
I notice that the time taken to get function is a lot higher compared to actually calling the function and getting the result back. Is there a way to load the function once and use it across multiple calls? 
Any help is appreciated, thanks :) 
As an extension to this, if anyone could point me to a repo with best practices for this kind of thing, I'd be really grateful. I am also unable to figure out how to use third party libraries in python code. For example, say I want to import from pandas, how do I do that?

**Comment by birkenfeld** [comment, marked as answer]
You can convert `sum_func` to a `Py<PyAny>` (just using `into()`).  That object can live outside the `with_gil` block, you can store it in a Rust struct or in a `lazy_static`/`once_cell` global.

As for using third party libraries, just do `PyModule::import(py, "pandas")` and go on from there.

**Reply by mejrs** [reply]
>  you can store it in a Rust struct or in a lazy_static/once_cell global.

We also have https://docs.rs/pyo3/latest/pyo3/once_cell/struct.GILOnceCell.html for that

**Reply by birkenfeld** [reply]
Ah yes, good to remember.

**Reply by vamshiaruru-virgodesigns** [reply]
Thank for your help! GILOnceCell seems to be working.
As for the third party imports, I was using poetry as my virtualenvironment manager, and looks like I have to explicitly set 
``` rust
    syspath.insert(0, "/<poetry_cache_path>/virtualenvs/axum-practice-x0KsYOAf-py3.9/lib/python3.9/site-packages").unwrap();
```
In my code. 
Is there a cleaner way than this?

**Reply by davidhewitt** [reply]
Ah the moment no, however there are some topics like #1741 (and same for windows) where we just need someone to figure out what a good universal solution could be.

**Reply by vamshiaruru-virgodesigns** [reply]
Understood, thanks!
